[ONNX] Trace model if quantization is detected

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76055
* #76002
* __->__ #75921
* #75920

Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

BowenBao committed:
[ONNX] Trace model if quantization is detected

Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

garymm commented on torch/onnx/utils.py:
```diff
@@ -495,6 +495,19 @@ def unpack_quantized_tensor(value):
+    This is because the quantized weights are stored in special `PackedParams` structure. They need to
+    be handled by `torch.jit.trace` to be exportable to ONNX.
+    """
+    if (any([hasattr(m, "_packed_params") for m in getattr(model, "modules", lambda: [])()]) or
```
I don't think you need a list inside `any`:
```suggestion
    if (any(hasattr(m, "_packed_params") for m in getattr(model, "modules", lambda: [])()) or
```

garymm commented on torch/onnx/utils.py:
```diff
@@ -495,6 +495,19 @@ def unpack_quantized_tensor(value):
+    be handled by `torch.jit.trace` to be exportable to ONNX.
+    """
+    if (any([hasattr(m, "_packed_params") for m in getattr(model, "modules", lambda: [])()]) or
+            any([getattr(arg, "is_quantized", False) for arg in args])):
```
```suggestion
            any(getattr(arg, "is_quantized", False) for arg in args)):
```

garymm commented on torch/onnx/utils.py:
```diff
@@ -573,7 +587,11 @@ def _model_to_graph(model, args, verbose=False,
+        # NOTE: assign_output_shape pass is not compatible with quantized outputs.
+        #       Quantized outputs are flattened to 3 values in ONNX, while packed as
+        #       single value in PyTorch.
+        if not any([getattr(out, "is_quantized", False) for out in output_tensors]):
```
ditto on list inside any

garymm commented on torch/onnx/utils.py:
```diff
@@ -573,7 +587,11 @@ def _model_to_graph(model, args, verbose=False,
 
         output_tensors, out_desc = torch._C._jit_flatten(tuple(output_wrapped))
-        torch._C._jit_pass_onnx_assign_output_shape(graph, output_tensors, out_desc, _onnx_shape_inference)
+        # NOTE: assign_output_shape pass is not compatible with quantized outputs.
```
Nit: Remove "NOTE:". The comment will be shorter and have the same informational content without it.

garymm commented on torch/onnx/utils.py:
```diff
@@ -495,6 +495,19 @@ def unpack_quantized_tensor(value):
+    original model.
+
+    This is because the quantized weights are stored in special `PackedParams` structure. They need to
+    be handled by `torch.jit.trace` to be exportable to ONNX.
```
Please add a link to an issue that we're using to track getting rid of the need for this.

I guess https://github.com/pytorch/pytorch/issues/47887 ?

BowenBao replied:
#47887 is a different issue. The closest tracking one is #75761. I will link it here and leave it open then.

BowenBao committed:
Update on "[ONNX] Trace model if quantization is detected"


Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge this
